### ISSUE TITLE
Openapi parsing performance improvements

### ISSUE THREAD

**Issue by natasha41575** [kubernetes-sigs/kustomize#4569]
OpenAPI parsing is a huge performance bottleneck for us. Related issues:
- https://github.com/kubernetes-sigs/kustomize/issues/4100
- https://github.com/kubernetes-sigs/kustomize/issues/2987
- https://github.com/kubernetes-sigs/kustomize/issues/3670

We discovered with these benchmarks that [parsing protobuf](https://github.com/kubernetes-sigs/kustomize/pull/4396) is much faster than parsing JSON OpenAPI schemas. kube-openapi now also supports a way to [convert from the `Gnostic v2` types to `spec.Swagger` types](https://github.com/kubernetes/kube-openapi/pull/283).

This means that instead of parsing the JSON to `spec.Swagger` types, which is what we do today, we can instead parse a proto schema to `Gnostic v2` types, and then convert the `Gnostic v2` to `spec.Swagger`.

Proto->gnostic-swagger takes ~39 ms.
Parsing JSON takes ~600 ms.

### Proposal

1. We should use protobuffer format instead of JSON for our builtin OpenAPI. https://github.com/kubernetes-sigs/kustomize/pull/4568 is a complete implementation of this.

2. Right now, if users have their own custom schema files, they provide them via the `openapi` field. However, the custom schema replaces the builtin schema entirely, so the user has to provide the entire schema themselves. Since they will typically provide the custom schema in JSON or YAML, they will not get the performance improvements of the builtin proto schema. To fix this, we can have the custom schema provided via the `openapi` field be supplementary to what is builtin, rather than replacing it completely. In this workflow, the users get two benefits:
  - they will only have to provide schemas for their custom types 
  - they will still get the huge performance boost of kustomize's builtin openapi in proto format even when they have custom resource schemas defined in JSON

3. We should add a `--proto` option to `kustomize openapi fetch` and also allow proto-formatted schemas in the `openapi` field. This would look something like:

```
openapi:
  path: my-proto.pb
```

**Comment by natasha41575** [comment]
cc @KnVerey

**Comment by shapirus** [comment]
Chiming in with my tests :)

Version 4.5.5 has much better performance than some older versions mentioned in previous tickets on this matter (5-7 times the speed, I would say, compared to versions around 3.9.4 or so), but it's still two times slower than 3.5.4. I've run a quick comparison on our manifests directory, containing a total of 460 kustomization.yaml files:

```
$ time (find . -type f -name kustomization.yaml -execdir kustomize-v3.5.4 build \; &>/dev/null)

real    0m16.606s
user    0m24.730s
sys     0m3.704s

$ time (find . -type f -name kustomization.yaml -execdir kustomize-v4.5.5 build \; &>/dev/null)

real    0m31.513s
user    0m58.568s
sys     0m7.938s
```

Just like before, it slows down when it encounters "unknown" resource kinds, such as SealedSecret, or actually any random string.

Interestingly enough, however, per my tests, v4.5.5 is about 2.5 times _faster_ than v3.5.4, if only known resource kinds are used: I performed a quick test on a simple set of manifests containing only a Deployment and a Service.

But once I add a SealedSecret there (still with v4.5.5), it becomes 5 times slower.

The 2.5x performance improvement over v3.5.4 is awesome, I wonder if it's possible to retain it for the case where arbitrary resource kinds are present as well.

**Comment by KnVerey** [comment]
/remove-lifecycle stale

We still need this part:

> We should add a --proto option to kustomize openapi fetch and also allow proto-formatted schemas in the openapi field.

**Comment by shapirus** [comment]
/remove-lifecycle stale

this won't just go away if this ticket is closed automatically.

**Comment by shapirus** [comment]
A gentle reminder that the issue is still there. In our case, this is a definite blocker preventing us from upgrading from approximately v3.5.4 to anything more recent.

v3.5.4, 17.8s:
```
$ time (find . -type f -name kustomization.yaml -execdir kustomize-v3.5.4 build \; &>/dev/null)

real    0m17.773s
user    0m26.537s
sys     0m5.160s
```

v5.0.0, the latest at the time of writing this, 28.5s, which is 1.6 times slower than v3.5.4:

```
$ time (find . -type f -name kustomization.yaml -execdir kustomize-v5.0.0 build \; &>/dev/null)

real    0m28.547s
user    0m53.858s
sys     0m6.810s
```

There was a big performance improvement at some point after v3.5.4 in processing the resources of "known" Kinds (e.g., Deployment, Service, etc.), and it is a pity that it remains useless, because it gets totally ruined by the poor performance of processing "unknown" Kinds such as e.g. SealedSecret.

**Comment by KnVerey** [comment]
@shapirus  thank you for posting your results with the lastest version
cc @ephesused -- perhaps this issue would interest you, as someone who has helped a lot with Kustomize performance recently

**Comment by ephesused** [comment]
Thanksk @KnVerey. I'm happy to take a look. I'll pick this up for investigation and if I don't think I'm in position to address it, I'll ask to hand it off.

@shapirus, if possible, will you please provide your test case or a simplified version that we can use for analysis? Thanks.

/assign

**Comment by shapirus** [comment]
Sure, I will, tomorrow.

**Comment by shapirus** [comment]
@ephesused here you go. Attached is a set of manifests that demonstrate the problem and a Dockerfile to build and run the tests conveniently. The list of versions to test is set in `Dockerfile:5`. It runs 300 iterations per each combination of kustomize versions and "known" (Service) and "unknown" (SealedSecret) kinds and measures the time it takes.

In a single command:

```docker run $(docker build -q .)```

[kustomize-performance-test.zip](https://github.com/kubernetes-sigs/kustomize/files/10755244/kustomize-performance-test.zip)

**Comment by ephesused** [comment]
Thanks, @shapirus. I hope to be able to start investigating early next week. I'll post here as I go.

**Comment by ephesused** [comment]
TL;DR: I plan to submit a PR to adjust something simple (gvk.ApiVersion) that should help. However, the PR does not address the issue of openapi parsing, so this issue will remain open past the initial PR. 

----

I dug around with some profiling to see what I could uncover.

As a start, it's worth noting that the test case is set up with only a single resource. That highlights the openapi parsing issue well, but (from what I can tell) the openapi parsing is a one-time expense. So the test case effectively is a worst-case scenario for the openapi side of things. That's fine, but I think it may put the focus on a misleading source of the issue.

To get execution in a range where I could profile with a little bit of confidence, I adjusted the test cases so they have 2000 resources (each identical to the original but with a different name).

The known full run uses a total of 570MB. The unknown full run uses a total of 1219MB, a difference of 649MB. The samples for memory were over 94% of the total in both cases, so the sampling is a good representation.

The known gvk is a v1 Secret, the unknown one is a bitnami.com/v1alpha1 SealedSecret.

The known run uses 121MB for resid.NewGVK. The unknown one uses 740MB for resid.NewGVK. That's a difference of 619MB, accounting for the vast majority of the difference for the full run.

For the known, all of the 121MB comes through strings.WriteString building the apiVersion ("v1").

For the unknown, 699MB of the 739MB comes through strings.WriteString building the apiVersion ("bitnami.com/v1alpha1"). This amount expands and contracts based on the number of resources. The other 41MB comes through openapi.initSchema. The openapi.initSchema amount doesn't seem to be affected by the resource count.

resource.CurId (resource.go line 447) calls resource.GetGvk (resource.go line 56) which calls resid.GvkFromNode (gvk.go line 31) which calls...

* resid.NewGVK (gvk.go line 24) which calls...
    * gvk.AsTypeMeta (gvk.go line 247) which calls...
        * gvk.ApiVersion (gvk.go line 124) which calls...
            * strings.WriteString
    * openapi.IsCertainlyClusterScoped (openapi.go line 406) which calls...
        * openapi.IsNamespaceScoped (openapi.go line 389) - the "v1" is recognized as a precomputed known type and returns; the "bitnami.com/v1alpha1" is not, which leads to a call into...
            * openapi.isNamespaceScopedFromSchema (openapi.go line 396) which calls...
                * openapi.initSchema (openapi.go line 625)

The known full run took 5.79s. The unknown full run took 7.05s, a difference of 1.26s. The sampling for cpu time was about 90% for both.

For what was sampled, the known run used 2.99s for resource.CurId. The unknown run used 3.85s for resource.CurId, a difference of 0.86s. The unknown run used 0.18s more for garbage collection (runtime.gcBcMarkWorker). The unknown run used 0.48s more for openapi.IsNamespaceScoped, 0.32s of which was openapi.initSchema.

Currently [gvk.ApiVersion uses strings.Builder](https://github.com/kubernetes-sigs/kustomize/blob/a502717/kyaml/resid/gvk.go#L123-L132) to create its output. Since the string construction is very simple, I don't think strings.Builder is the best approach. When I created a benchmark to test its current execution, here's what happens for bitnami.com/v1alpha1:

```
BenchmarkApiVersion/5-8         16658961                67.41 ns/op           48 B/op          2 allocs/op
```

With the following gvk.ApiVersion code, the execution runs in 1/3 the time and avoids allocations:

```go
	if x.Group != "" {
		return x.Group + "/" + x.Version
	}
	return x.Version
```

```
BenchmarkApiVersion/5-8         57177163                21.71 ns/op            0 B/op          0 allocs/op
```

So my thinking is that the first step is to adjust gvk.ApiVersion. That should make its execution quicker and should reduce load on garbage collection - which should help overall runtime.

## Profile results

### Known use case, CPU
![known use case cpu](https://user-images.githubusercontent.com/11396029/220238648-9e5e73a8-9b71-4729-b45c-8070deda230f.svg)

### Unknown use case, CPU
![unknown use case cpu](https://user-images.githubusercontent.com/11396029/220238646-d758775d-3d1e-4c22-bffd-a45b463d268f.svg)

### Known use case, memory
![known use case memory](https://user-images.githubusercontent.com/11396029/220238645-de82fe2a-c362-4932-9d92-0214071c27dd.svg)

### Unknown use case, memory
![unknown use case memory](https://user-images.githubusercontent.com/11396029/220238647-e048e82a-56cc-4d18-a654-792e550a6c12.svg)

**Comment by shapirus** [comment]
A note regarding this:

```it's worth noting that the test case is set up with only a single resource.```

it is actually a close representation of our use case: hundreds of small sets of resources, each having its own `kustomization.yaml` (microservice architecture where each microservice is additionally split into variants) for each of which a new invocation of `kustomize` is done. This is why this performance degradation became so very well noticeable in our situation.

**Comment by ephesused** [comment]
I follow, @shapirus. To be clear, I don't intend to stop with the initial PR. That memory use jumped out at me when looking at the profile results, and it's a quick thing to address. I figured I might as well get that in place, with the hopes that it might improve things on your end at least somewhat.

**Comment by ephesused** [comment]
FWIW... In my testing with @shapirus's test scenario, #5808 yields about a 10% runtime improvement over v5.0.0 for both known and unknown resources.

**Comment by ephesused** [comment]
Sorry, other commitments have claimed my time recently. I did do some further analysis, but haven't had time to dig and come up with ideas for improvement. I hope to have more time later this week.

I set up a test case that ran 100 executions of the one-resource known & unknown test cases (resetting the global "already initialized" flag each time, to mimic the user experience of running kustomize fresh each time). Below, kn stands for known and uk stands for unknown.

The cpu run-time results for unknown resources are more than three times that of the known resources. That's meaningfully worse than what @shapirus noted previously in this issue. The memory differences might look surprising at first, but they are what I expected based on earlier analysis (each openapi initialization requires a little more than 40MB).

```
CPU
kn-cpu-master-976193ce7.svg: Duration: 2.22s, Total samples = 1.51s (67.88%)
uk-cpu-master-976193ce7.svg: Duration: 7.25s, Total samples = 6.76s (93.26%)

Memory
kn-mem-master-976193ce7.svg: Showing nodes accounting for 96.96MB, 65.62% of 147.76MB total
uk-mem-master-976193ce7.svg: Showing nodes accounting for 3.92GB, 88.04% of 4.45GB total
```

![known resource cpu usage](https://user-images.githubusercontent.com/11396029/221879835-d1bd4492-3f67-4e4a-9cea-661daf13370e.svg)
![unknown resource cpu usage](https://user-images.githubusercontent.com/11396029/221879839-da1b217f-9d76-4126-86cc-6849ed68ee3c.svg)
![known resource memory usage](https://user-images.githubusercontent.com/11396029/221879838-9b34bff7-9f0f-4ef3-b6e4-82385cb25fdd.svg)
![unknown resource memory usage](https://user-images.githubusercontent.com/11396029/221879833-b5b22962-48be-4ef9-9cf3-febff175b628.svg)

**Comment by ephesused** [comment]
After digging around, I have a possible approach in mind. The work done in #4152 ensures the precomputed namespace scoped information aligns with the default schema. The implication is that anything not found in the precomputed information is not in the default schema. Knowing that, there's no point in loading the default schema when trying to determine namespace scoping. In that case, if it's not precomputed, we know that the typeMeta won't be found and that openapi.isNamespaceScopedFromSchema must return false, false.

So the idea is to put a short-circuit in isNamespaceScopedFromSchema when the schema is the default.

**Comment by ephesused** [comment]
Just a quick note to say that while I expect this PR to be small, I haven't been able to carve the time to do the coding. I intend to do so within the next few days.

**Comment by shapirus** [comment]
...lots of failing checks on that PR :)

**Comment by shapirus** [comment]
@ephesused since PR https://github.com/kubernetes-sigs/kustomize/pull/5076 has finally been merged, I decided to run my tests again. Unfortunately, there has been no release with that PR yet, so I had to build current master branch (2f99707792228c921e2d477dbad7fc4898691365) to test it.

It looks awesome. Not only the slow parsing of unknown kinds is fixed, but it is now 4.7 times faster than kustomize v3.5.4 (the release that I'm still stuck with because of those performance issues) _for both known and unknown kinds_!

(this is all x86_64, haven't tested it on arm64 yet)

```
################# STARTING TESTS
# kustomize versions: v3.5.4 v5.1.1 2f99707792228c921e2d477dbad7fc4898691365

## testing known resource kinds ##
kustomize v3.5.4, 1000 iterations
real    0m21.163s

kustomize v5.1.1, 1000 iterations
real    0m4.641s

kustomize 2f99707792228c921e2d477dbad7fc4898691365, 1000 iterations
real    0m4.468s


## testing unknown resource kinds ##
kustomize v3.5.4, 1000 iterations
real    0m21.101s

kustomize v5.1.1, 1000 iterations
real    0m57.450s

kustomize 2f99707792228c921e2d477dbad7fc4898691365, 1000 iterations
real    0m4.455s

```

As far as my use case is concerned, I consider the issue resolved at this point, thank you very much @ephesused.

**Comment by shapirus** [comment]
5.2.0 has just been released, and I can confirm that it is still as fast as the revision I tested in https://github.com/kubernetes-sigs/kustomize/issues/4569#issuecomment-1762786077.
